src/managers/balancer.py

```diff
@@ -80,3 +181,114 @@ def create_internal_topics(self) -> None:
                     ],
                 )
                 logger.info(f"Created topic {topic}")
+
+    def rebalance(self, mode: str, dryrun: bool = True) -> requests.Response:
+        """Triggers a full Kafka cluster partition rebalance."""
+        rebalance_request = self.cruise_control.post(
+            "rebalance", dryrun=dryrun
+        )  # FIXME: will need updating to include add/remove_broker when it works
+        user_task_id = rebalance_request.headers.get("User-Task-ID", "")
+
+        # block entire charm event handling while rebalance in progress
+        while (
+            "Completed" not in self.cruise_control.get_task_status(user_task_id=user_task_id)
+            or self.cruise_control.executing
+        ):
+            logger.info(
+                f"Waiting for '{mode}' rebalance {'dryrun' if dryrun else ''} task to complete..."
+            )
+            time.sleep(10)  # sleep needed as CC API rejects too many requests within a short time
+
+        logger.info(
+            f"'{mode}' rebalance task finished - {self.cruise_control.get_task_status(user_task_id=user_task_id)}"
+        )
+        return rebalance_request
+
+    def _get_storage_size(self, path: str) -> int:
+        """Gets the total storage volume of a mounted filepath, in KB."""
+        return int(self.dependent.workload.exec(f"df --output=size {path} | sed 1d"))
+
+    def _build_new_key(self, nested_key: str, nested_value: JSON) -> dict[str, JSON]:
+        """Builds a nested key:value pair for JSON lists from the output of a rebalance proposal.
+
+        The keys where this is needed are `brokers`, `hosts` and `goalSummary` goals. Turns this:
+        ```
+        "loadAfterOptimization": {
+            "brokers": [
+              {
+                "BrokerState": "ALIVE",
+                "Broker": 0,
+              },
+              {
+                "BrokerState": "ALIVE",
+                "Broker": 1,
+              },
+        }
+        ```
+
+        into this:
+
+        ```
+        "loadAfterOptimization": {
+            "brokers.0":
+              {
+                "brokerstate": "ALIVE",
+                "broker": 0,
+              },
+            "brokers.1":
+              {
+                "brokerstate": "ALIVE",
+                "broker": 1,
+              },
+        }
```

Comment:
**praise**: having an example in the docstring is so helpful to understand the code

Reply:
Yeah! I thought it was far too complicated to not explicitly say what it was doing, given the messy JSON recursion.